### PR TITLE
material-ui: Fix IconMenu properties.

### PR DESCRIPTION
**Pull request by mtgto**
Fix the failing #22404 .

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/mui-org/material-ui/blob/v0.20.0/src/IconMenu/IconMenu.js
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

----

I tried to run `tsc` successful at locally. But Travis CI failed because `Error: Error: stdout maxBuffer exceeded`.
https://travis-ci.org/mtgto/DefinitelyTyped/jobs/324934573

I don't know how to avoid the CI failure.
I also read this document, but it doesn't help me.
http://definitelytyped.org/guides/pull-request.html#travis-ci-for-your-dt-fork

### CODE REVIEW

**Comment by mhegazy** [comment]
@theigor @DaIgeb @tkrotoff @nbgraham @MrCeleryman @mthmulders @Harurow @estaub @devuxer @minodisk @cauld @allienna @alitaheri @dan-j @weronka @vincealdrin @RaulTsc @timothykang @tqn @syamatoo @schlesingermatthias @ospfranco @onigoetz @MazeChaZer @johnnyreilly can you please review this change

File: types/material-ui/index.d.ts [hunk 1]
```diff
@@ -1241,37 +1241,37 @@ declare namespace __MaterialUI {
...
+            animated?: boolean;
             animation?: React.ComponentClass<Popover.PopoverAnimationProps>;
             className?: string;
-            iconButtonElement: React.ReactElement<IconButtonProps>;
+            clickCloseDelay?: number;
+            iconButtonElement: React.ReactElement<IconButton>;
             iconStyle?: React.CSSProperties;
             menuStyle?: React.CSSProperties;
-            onItemTouchTap?(e: React.SyntheticEvent<{}>, item: MenuItem): void;
+            onClick?(item: IconButton): void;
```
Comment (nbgraham) [inline types/material-ui/index.d.ts]: The signature for `onClick` on the [docs](http://www.material-ui.com/#/components/icon-menu) is:  
`function(event: object) => void`  
*event*: Click event targeting the `IconButton` element.  

So I think this should be:  
`onClick?(e: React.SyntheticEvent<{}>): void;`

Comment (mtgto) [reply types/material-ui/index.d.ts]: @nbgraham You are right. I mistake to read the docs (the argument of `onClick` is `IconButton`).

**Review by dan-j — APPROVED** [review]

File: types/material-ui/index.d.ts [hunk 2]
```diff
@@ -1241,37 +1241,37 @@ declare namespace __MaterialUI {
         export interface IconMenuProps {
             // <Menu/> is the element that get the 'other' properties
             anchorOrigin?: propTypes.origin;
+            animated?: boolean;
             animation?: React.ComponentClass<Popover.PopoverAnimationProps>;
             className?: string;
-            iconButtonElement: React.ReactElement<IconButtonProps>;
+            clickCloseDelay?: number;
+            iconButtonElement: React.ReactElement<IconButton>;
```
Comment (andrioli) [inline types/material-ui/index.d.ts]: Is this line change correct? After update my types I got an error

```
(19,9): error TS2322: Type '{ iconButtonElement: Element; children: (false | Element)[]; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<IconMenu> & Readonly<{ children?: ReactNode; }> & ...'.
  Type '{ iconButtonElement: Element; children: (false | Element)[]; }' is not assignable to type 'Readonly<IconMenuProps>'.
    Types of property 'children' are incompatible.
      Type '(false | Element)[]' is not assignable to type 'ReactElement<MenuItemProps> | ReactElement<MenuItemProps>[] | undefined'.
        Type '(false | Element)[]' is not assignable to type 'ReactElement<MenuItemProps>[]'.
          Type 'false | Element' is not assignable to type 'ReactElement<MenuItemProps>'.
            Type 'false' is not assignable to type 'ReactElement<MenuItemProps>'.
```

File: types/material-ui/index.d.ts [hunk 3]
```diff
@@ -1241,37 +1241,37 @@ declare namespace __MaterialUI {
...
             open?: boolean;
-            style?: React.CSSProperties;
             targetOrigin?: propTypes.origin;
-            clickCloseDelay?: number;
             useLayerForClickAway?: boolean;
 
-            animated?: boolean;
+            // Other properties from <Menu/>
             autoWidth?: boolean;
+            children?: Array<React.ReactElement<MenuItemProps>>;
```
Comment (andrioli) [inline types/material-ui/index.d.ts]: Just found the bug from my previous [comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/22663/files#r165432206).

The problem is that I'm adding MenuItem conditionally. Example:
```
<MenuItem ... />
{this.props.someCondition &&
    <MenuItem ... />
}
<MenuItem ... />
```

So in my case here the type must be `Array<false|React.ReactElement<MenuItemProps>>`

Comment (mtgto) [reply types/material-ui/index.d.ts]: @andrioli
I found the bug in MenuItem's children, so fixed #22910 .
Your bug is fixed this PR, or not yet fixed?
(what means `Array<false>` ?)

Can you show the sample code which causes the error?

Comment (andrioli) [reply types/material-ui/index.d.ts]: No, is not yet fixed.

If you try to create a `IconMenu` with the items from my previous comment the compile will break.

Complete Example

```
import IconButton from 'material-ui/IconButton';
import IconMenu from 'material-ui/IconMenu';
import MenuItem from 'material-ui/MenuItem';
import PersonIcon from 'material-ui/svg-icons/social/person';
import * as React from 'react';

interface IconMenuBugProps {
  someCondition: boolean;
}

class IconMenuBug extends React.Component<IconMenuBugProps> {

  public render() {
    return (
      <IconMenu
        iconButtonElement={<IconButton><PersonIcon color="#AD3489" /></IconButton>}
      >
        <MenuItem value="0" primaryText="Hey" />
        {/* may be false */}
        {!this.props.someCondition &&
          <MenuItem value="1" primaryText="Ho" />
        }
        {/* may be false */}
        {this.props.someCondition &&
          <MenuItem value="2" primaryText="Let's Go" />
        }
      </IconMenu>
    );
  }

}
```

Note that the children array can also contains `false` values other than only `React.ReactElement<MenuItemProps>`

Comment (mtgto) [reply types/material-ui/index.d.ts]: Thank you, I understand.
I created #23387.

Sorry about any inconvenience.

Comment (andrioli) [reply types/material-ui/index.d.ts]: Thank you for the PR

**Pull request merged**